src/mono/mono/component/mini-wasi-debugger.c

```diff
@@ -0,0 +1,172 @@
+__attribute__((import_name("sock_accept")))
+int sock_accept(int fd, int fdflags, int* result_ptr);
+
+static long long
+timeInMilliseconds ()
+{
+	struct timeval tv;
+	gettimeofday(&tv,NULL);
+	return (((long long)tv.tv_sec)*1000)+(tv.tv_usec/1000);
+}
+
+static int
+wasi_transport_recv (void *buf, int len)
+{
+	int res;
+	int total = 0;
+	int fd = conn_fd;
+	int again = 0;
+
+	do {
+		res = read (fd, (char *) buf + total, len - total);
+		if (res > 0)
+			total += res;
+		again++;
+		if ((res > 0 && total < len) || (res == -1 && again < retry_receive_message)) {
+			// Wasmtime on Windows doesn't seem to be able to sleep for short periods like 1ms so we'll have to spinlock
+			long long start = timeInMilliseconds ();
+			while (timeInMilliseconds () < start + (connection_wait/1000));
+		} else {
+			break;
+		}
+	} while (true);
+	return total;
+}
+
+static gboolean
+wasi_transport_send (void *data, int len)
+{
+	int res;
+	do {
+		res = write (conn_fd, (const char*)data, len);
+	} while (res == -1);
+
+	return res == len;
+}
+
+/*
+ * wasi_transport_connect:
+ *
+ *   listen on the socket that is already opened by wasm runtime
+ */
+static void
+wasi_transport_connect (const char *socket_fd)
+{
+	PRINT_DEBUG_MSG (1, "wasi_transport_connect.\n", conn_fd);
+	bool handshake_ok = FALSE;
+	
+	while (!handshake_ok) {
+		PRINT_DEBUG_MSG (1, "Waiting for connection from client, socket fd=%s.\n", socket_fd);
+		sock_accept (atoi(socket_fd), 4, &conn_fd);
+		int res = -1;
+		if (conn_fd != -1) {
+			res = write (conn_fd, (const char*)"", 0);
+			if (res != -1) {
+				handshake_ok = mono_component_debugger ()->transport_handshake ();
+			}
+		}
+		if (conn_fd == -1 || res == -1) {
+			sleep(1);
+			continue;
+		}
+	}
+	PRINT_DEBUG_MSG (1, "Accepted connection from client, socket fd=%d.\n", conn_fd);
+}
+
+static void
+wasi_transport_close1 (void)
+{
+/*	shutdown (conn_fd, SHUT_RD);
+	shutdown (listen_fd, SHUT_RDWR);
+	close (listen_fd);*/
+}
+
+static void
+wasi_transport_close2 (void)
+{
+//	shutdown (conn_fd, SHUT_RDWR);
+}
+
+
+static void 
+mono_wasi_start_debugger_thread (MonoError *error)
+{
+	mono_debugger_agent_receive_and_process_command ();
+	connection_wait = 250;
+}
+
+static void
+mono_wasi_suspend_vm (void)
+{
+
+}
+
+static void
+mono_wasi_debugger_init (void)
+{
+	DebuggerTransport trans;
+	trans.name = "wasi_socket";
+	trans.send = wasi_transport_send;
+	trans.connect = wasi_transport_connect;
+	trans.recv = wasi_transport_recv;
+	trans.send = wasi_transport_send;
+	trans.close1 = wasi_transport_close1;
+
+	mono_debugger_agent_register_transport (&trans);
+
+	mono_debugger_agent_init_internal();
+	
+	mono_debugger_agent_initialize_function_pointers(mono_wasi_start_debugger_thread, mono_wasi_suspend_vm, mono_wasi_suspend_current);
+}
+
+static void 
+mono_wasi_receive_and_process_command_from_debugger_agent (void)
+{
+	retry_receive_message = 2;
+	mono_debugger_agent_receive_and_process_command ();
+	retry_receive_message = 50;
```

Comment:
when it comes from interpreter we just want to check if there is any breakpoint inserted for example, so we only try to get message 2 times.
when it comes from debugger-agent, something like we are already paused on a step then we are waiting for a resume or something else, we retry more because we don't care about performance.

Reply:
Could you please add this as a comment in the code?